backend/asmgen.ml

```diff
@@ -233,37 +237,61 @@ let count_duplicate_spills_reloads_in_block (cfg_with_infos : Cfg_with_infos.t)
       dup_spills, new_dup_reloads, seen_spill_regs, Reg.Set.add reg seen_reload_regs
     | _ -> acc
   in
-  let count_per_block _ (block : Cfg.basic_block) (dup_spills, dup_reloads) =
-    let dup_spills, dup_reloads, _, _ =
-      DLL.fold_left block.body ~f:count_per_inst
-        ~init:(dup_spills, dup_reloads, Reg.Set.empty, Reg.Set.empty)
-    in
-    dup_spills, dup_reloads
+  let dup_spills, dup_reloads, _, _ =
+    DLL.fold_left block.body ~f:count_per_inst
+      ~init:(0, 0, Reg.Set.empty, Reg.Set.empty)
   in
-  Cfg_with_infos.fold_blocks cfg_with_infos ~f:count_per_block ~init:(0, 0)
+  dup_spills, dup_reloads
 
-let count_spills_reloads (cfg_with_infos : Cfg_with_infos.t) =
+let count_spills_reloads (block : Cfg.basic_block) =
   let f ((spills, reloads) as acc) (instr : Cfg.basic Cfg.instruction) =
-  match instr.desc with
-  | Op Spill -> (spills + 1, reloads)
-  | Op Reload -> (spills, reloads + 1)
-  | _ -> acc
+    match instr.desc with
+    | Op Spill -> spills + 1, reloads
+    | Op Reload -> spills, reloads + 1
+    | _ -> acc
   in
-  Cfg_with_infos.fold_body_instructions cfg_with_infos ~f ~init:(0, 0)
-
-let regalloc_profile =
-  let counter_f cfg =
-    let dup_spills, dup_reloads = count_duplicate_spills_reloads_in_block cfg in
-    let spills, reloads = count_spills_reloads cfg in
-    Profile.Counters.create ()
-    |> Profile.Counters.set "block_duplicate_spill" dup_spills
-    |> Profile.Counters.set "block_duplicate_reload" dup_reloads
-    |> Profile.Counters.set "spill" spills
-    |> Profile.Counters.set "reload" reloads
+  DLL.fold_left ~f ~init:(0, 0) block.body
+
+let cfg_block_counters block =
+  let dup_spills, dup_reloads = count_duplicate_spills_reloads_in_block block in
+  let spills, reloads = count_spills_reloads block in
+  Profile.Counters.create ()
+  |> Profile.Counters.set "block_duplicate_spill" dup_spills
+  |> Profile.Counters.set "block_duplicate_reload" dup_reloads
+  |> Profile.Counters.set "spill" spills
+  |> Profile.Counters.set "reload" reloads
+
+let whole_cfg_counters _ = Profile.Counters.create ()
```

Comment:
This catch-all pattern looks a bit
suspicious; should it be `()`?

Reply:
I have left this in there to allow anyone in the future to add counters that require the whole CFG to produce a count. I have added a type annotation and comments to both `cfg_block_counters` and `whole_cfg_counters`.

Reply:
thanks